cocotb/triggers.py

```diff
@@ -279,89 +314,138 @@ def FallingEdge(signal):
 
 
-class ClockCycles(_Edge):
+class Sequential(PythonTrigger):
     """
-    Execution will resume after N rising edges or N falling edges
+    Combines multiple triggers together sequentially.  Coroutine will continue when all
+    triggers have fired, each in turn.
     """
-    def __init__(self, signal, num_cycles, rising=True):
-        _Edge.__init__(self, signal)
-        self.num_cycles = num_cycles
-        if rising is True:
-            self._rising = 1
-        else:
-            self._rising = 2
+    def __init__(self, *args):
+        PythonTrigger.__init__(self)
+        self._triggers = list(args)
+        self._active = None
+        self._fired = []
+        self._gpi_based = False
+        try:
+            for trigger in self._triggers:
+                if not isinstance(trigger, Trigger):
+                    raise TriggerException("All combined triggers must be "
+                                           "instances of Trigger! Got: %s" %
+                                           trigger.__class__.__name__)
+                self._gpi_based |= trigger.needs_advance()
+        except Exception:
+            raise TriggerException("%s requires a list of Trigger objects" %
+                                   self.__class__.__name__)
 
     def prime(self, callback):
-        self._callback = callback
+        if callback not in self._callback_list:
+            self._callback_list.append(callback)
+        if not self._triggers:
+            raise TriggerException("%s requires a list of Trigger objects, but they vanished" %
+                                   self.__class__.__name__)
+        self._active = self._triggers.pop(0)
+        Trigger.prime(self)
+        self._active.prime(self._check_all_fired)
 
-        def _check(obj):
-            self.unprime()
+    def _check_all_fired(self, trigger):
+        self._fired.append(trigger)
+        if len(self._triggers):
+            self._active = self._triggers.pop(0)
+            self._active.prime(self._check_all_fired)
+        else:
+            # reset lists in case we are primed again
+            self._triggers = self._fired
+            self._fired = []
+            self._callback(self)
 
-            if self.signal.value:
-                self.num_cycles -= 1
+    def needs_advance(self):
+        return self._gpi_based
 
-                if self.num_cycles <= 0:
-                    self._callback(self)
-                    return
 
-            self.cbhdl = simulator.register_value_change_callback(self.signal.
-                                                                  _handle,
-                                                                  _check,
-                                                                  self._rising,
-                                                                  self)
-            if self.cbhdl == 0:
-                raise_error(self, "Unable set up %s Trigger" % (str(self)))
+class RepeatNTriggers(PythonTrigger):
```

Comment:
What is the advantage of this over just
```
@cocotb.coroutine
def repeat_n_trigger(trig, n):
    for i in range(n):
        yield trig
```
It seems to me that you only need two layers here - a trigger capable of calling one callback, and an event loop capable of dispatching to arbitrary many coroutines. Heirachical triggers just duplicate this dispatch logic in two places.

Reply:
> Are you suggesting that it no longer be a trigger?

Yes. Either it's a primitive trigger with exactly one callback, or it's a compound coroutine that is dispatched to by triggers in the scheduler

Reply:
> Yes. Either it's a primitive trigger with exactly one callback, or it's a compound coroutine that is dispatched to by triggers in the scheduler

I think this would be a bad idea. There are at least a few places in cocotb where triggers and coroutines do different things or behave differently (one of them is fixed by #618, but I suspect there are more places). Changing "ClockCycles" to be a coroutine when it has been a trigger up until now will probably break at least some existing testbenches. I'm not sure that it is worth doing that.

But I shall leave the decision of which approach to adopt to the maintainers.

Reply:
Note that it's possible to make `ClockCycles()` return a `_Join()` trigger, as I do [here](https://github.com/potentialventures/cocotb/issues/520#issuecomment-411667489). In my mind, compatibility is the only reason to keep `ClockCycles` as a trigger, and I'd rather that was done without introducing a bunch of redundant scheduling models.

Reply:
Fair enough. I mean, I do also think "triggers" (the things currently called triggers in ```cocotb.triggers ```) and coroutines *should* be interchangeable to the end-user (more or less), which is why I proposed #618.

I am only attached to this solution because it has already been implemented and seems to work (and @stuarthodgson said he was planning to merge it in the original PR's discussion). If you want to open a separate PR for your proposal, that would probably be valuable too.

Reply:
I'm holding out on #640 before doing that, since that reworks how triggers and coroutines produce results